src/resolver/manifest.cc

```diff
@@ -0,0 +1,533 @@
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+static void
+from_json( const nlohmann::json & jfrom, ManifestRaw::Options::Allows & allow )
+{
+  if ( ! jfrom.is_object() )
+    {
+      std::string aOrAn = jfrom.is_array() ? " an " : " a ";
+      throw InvalidManifestFileException(
+        "Manifest field `options.allow' must be an object, but is" + aOrAn
+        + std::string( jfrom.type_name() ) + '.' );
+    }
+
+  /* Clear fields. */
+  allow.licenses = std::nullopt;
+  allow.unfree   = std::nullopt;
+  allow.broken   = std::nullopt;
+
+  for ( const auto & [key, value] : jfrom.items() )
+    {
+      if ( key == "unfree" )
+        {
+          try
+            {
+              value.get_to( allow.unfree );
+            }
+          catch ( const nlohmann::json::exception & )
+            {
+              throw InvalidManifestFileException(
+                "Failed to parse manifest field `options.allow.unfree' "
+                "with value: "
+                + value.dump() );
+            }
+        }
+      else if ( key == "broken" )
+        {
+          try
+            {
+              value.get_to( allow.broken );
+            }
+          catch ( const nlohmann::json::exception & )
+            {
+              throw InvalidManifestFileException(
+                "Failed to parse manifest field `options.allow.broken' "
+                "with value: "
+                + value.dump() );
+            }
+        }
+      else if ( key == "licenses" )
+        {
+          try
+            {
+              value.get_to( allow.licenses );
+            }
+          catch ( const nlohmann::json::exception & )
+            {
+              throw InvalidManifestFileException(
+                "Failed to parse manifest field `options.allow.licenses' "
+                "with value: "
+                + value.dump() );
+            }
+        }
+      else
+        {
+          throw InvalidManifestFileException(
+            "Unrecognized manifest field `env-base.options.allow." + key
+            + "'." );
+        }
+    }
+}
+
+
+static void
+to_json( nlohmann::json & jto, const ManifestRaw::Options::Allows & allow )
+{
+  if ( allow.unfree.has_value() ) { jto = { { "unfree", *allow.unfree } }; }
+  else { jto = nlohmann::json::object(); }
+
+  if ( allow.broken.has_value() ) { jto.emplace( "broken", *allow.broken ); }
+  if ( allow.licenses.has_value() )
+    {
+      jto.emplace( "licenses", *allow.licenses );
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+void
+from_json( const nlohmann::json & jfrom, ManifestRaw::Options & opts )
+{
+  if ( ! jfrom.is_object() )
+    {
+      std::string aOrAn = jfrom.is_array() ? " an " : " a ";
+      throw InvalidManifestFileException(
+        "Manifest field `options' must be an object, but is" + aOrAn
+        + std::string( jfrom.type_name() ) + '.' );
+    }
+
```

Comment:
question: I don't quite understand why you're clearing fields in the other `from_json` methods (are we expecting we'll be overwriting an existing `ManifestRaw` object?) but are you deliberately skipping doing that here?

Reply:
I clear using the sub-parsers instead of trying to clear the whole record at the top level.

In many cases I prefer an explicit `foo.clear()` function for this; but since a lot of these structs are only used here I just cleared them in the `from_json` routines.